persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java

```diff
@@ -98,37 +117,186 @@ public void beforeChange(
+      notifyDown(endpoint);
+    }
+  }
 
-  private void addStargateNode(InetAddressAndPort endpoint, EndpointState epState) {
-    if (!isStargateNode(epState) || !liveStargateNodes.add(endpoint)) {
+  private void notifyUp(InetAddressAndPort endpoint) {
+    if (!StorageService.instance.isRpcReady(endpoint) || !Gossiper.instance.isAlive(endpoint)) {
       return;
     }
 
-    for (IEndpointLifecycleSubscriber subscriber : subscribers) subscriber.onJoinCluster(endpoint);
+    if (endpointsPendingJoinedNotification.remove(endpoint)) {
+      notifyJoinCluster(endpoint);
+    }
+
+    InetAddressAndPort nativeAddress = getNativeAddress(endpoint);
+    for (EventListener listener : listeners) {
+      listener.onUp(nativeAddress.address, nativeAddress.port);
+    }
+  }
+
+  private void applyState(
+      InetAddressAndPort endpoint,
+      ApplicationState state,
+      VersionedValue value,
+      EndpointState epState) {
+    switch (state) {
+      case RELEASE_VERSION:
+        StargateSystemKeyspace.updatePeerInfo(endpoint, "release_version", value.value);
+        break;
+      case DC:
+        StargateSystemKeyspace.updatePeerInfo(endpoint, "data_center", value.value);
+        break;
+      case RACK:
+        StargateSystemKeyspace.updatePeerInfo(endpoint, "rack", value.value);
+        break;
+      case RPC_ADDRESS:
+        try {
+          StargateSystemKeyspace.updatePeerInfo(
+              endpoint, "rpc_address", InetAddress.getByName(value.value));
+        } catch (UnknownHostException e) {
+          throw new RuntimeException(e);
+        }
+        break;
+      case NATIVE_ADDRESS_AND_PORT:
+        try {
+          InetAddressAndPort address = InetAddressAndPort.getByName(value.value);
+          StargateSystemKeyspace.updatePeerNativeAddress(endpoint, address);
+        } catch (UnknownHostException e) {
+          throw new RuntimeException(e);
+        }
+        break;
+      case SCHEMA:
+        // Use a fix schema version for all peers (always in agreement) because stargate waits
+        // for DDL queries to reach agreement before returning.
+        StargateSystemKeyspace.updatePeerInfo(
+            endpoint, "schema_version", StargateSystemKeyspace.SCHEMA_VERSION);
+
+        // This fix schedules a schema pull for the non-member node and is required because
+        // `StorageService.onChange()` doesn't do this for non-member nodes.
+        MigrationManager.instance.scheduleSchemaPull(endpoint, epState);
+        break;
+      case HOST_ID:
+        StargateSystemKeyspace.updatePeerInfo(endpoint, "host_id", UUID.fromString(value.value));
+        break;
+      case RPC_READY:
+        notifyRpcChange(endpoint, epState.isRpcReady());
+        break;
+    }
+  }
+
+  private void updateTokens(InetAddressAndPort endpoint) {
+    StargateSystemKeyspace.updatePeerInfo(
+        endpoint,
+        "tokens",
+        Collections.singleton(DatabaseDescriptor.getPartitioner().getMinimumToken().toString()));
```

Comment:
is it safe to use `toString()`? Maybe we should have a dedicated method for parsing to string?

Reply:
You're right, I'm now using `TokenFactory.toString(Token)` which is the correct way of doing this.